Increase Test Coverage #87

th1nkNXT opened the pull request:
Hi there.
I only added one test case to get some feedback first.

- I added Lombok to reduce boilerplate code (identical to my changes in #370)
- I introduced the builder pattern to improve code readability for the Block class

Would be good if you can me give some advice about creating a valid block in the test case and how to simulate this correctly (just took some data from another test). If this is finished i can create more tests e.g. like verifying the reward account.
If there is something wrong or should be improved just give me a message.

I am motivated to help, so feel free to contact me if i can help you also on other tasks.
If you want to know something about my coding experiences and background you can also give me a message. Thanks.

jjos2372 commented:
Thank you, we are working on a lot of fronts right now. I'll check this out when the re-brand activities are close to be finished.

ohager commented on src/brs/db/sql/SqlBlockDb.java:
```diff
@@ -102,9 +103,26 @@ public Block loadBlock(BlockRecord r) throws BurstException.ValidationException
-    return new Block(version, timestamp, previousBlockId, totalAmountNQT, totalFeeNQT, payloadLength, payloadHash,
-                         generatorPublicKey, generationSignature, blockSignature, previousBlockHash,
-                         cumulativeDifficulty, baseTarget, nextBlockId, height, id, nonce, blockATs);
+    return Block.builder()
```
Is this relevant for unit tests, or "just" a useful refactoring?

th1nkNXT replied:
I would say it is a useful refactoring because we are more flexible when creating a new Block (which can help when creating test cases which do not need every attribute to be filled). Furthermore it increases readability IMHO.

th1nkNXT replied:
I just wanted to show a possibility which Lombok offers. It increases readibility as you can see at first sight, which param is set to which value. You are right, that the current state in too generic and allows the creation of "invalid" block objects. 

From my point of view the existing code needs to be improved because the readability and documentation is not so good. If you provide different constructors there should be at least a JavaDoc when to use which constructor. A better way to go would be to just use a package private constructor (so that no one can access from a "higher" level) and use the factory pattern to create complex objects / separate the different types of creation. 

In the old code there were 2 constructors. Could you provide some more information for the different creation approaches? Maybe it could be also reduced to 1? Otherwise my suggestion would be to provide different factory classes / methods to create a block for the different use case. If you think that would be a good idea, i can write the needed classes/methods. If you have another idea in mind just give me an example.

ohager reviewed (approved):
LGTM - but @jjos2372 has final words :smile:

ohager commented:
old and stale.... shall we close this? Pity for the efforts done here.

th1nkNXT commented:
@ohager @jjos2372 
So, i am back :-)
To make it more easy for you and to avoid losing time on general architecture discussions:
- i updated my branch based on develop (Java 11, gradle, ..)
- i removed Lombok usages

Test is green.

frankTheTank72 commented:
@ohager can we merge this ?

frankTheTank72 commented:
@jjos2372 @ohager   what should happen witth this PR ?

ohager commented:
> @jjos2372 @ohager what should happen witth this PR ?

I'm a bit concerned that we have so long pending contributions here...or we declare stale and close this, or we (as having no conflicts) we merge it.

ohager commented:
Close due to stale. Hurts me a bit to see a PR pending for so much time, and finally not accepting a community members contribution... but bigger changes are coming in.
